Add Tutorial for how to connect a Docker Compose workflow to Kubernetes

ezodude opened the pull request:
Resolves https://github.com/appvia/kev/issues/203

Assumes the theme of connecting a Docker Compose workflow to Kubernetes.

It's based on the problems developers encounter when migrating to Kubernetes. This was researched [using these articles](https://www.google.com/search?hl=en&q=migrating%20a%20docker%20compose%20workflow%20to%20kubernetes).

Contains,
- A new `Installation` section.
- A refactored `Quickstart` section.
- A new `Tutorial` section.

marcinc commented on README.md:
```diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
+- `kev render` - detects, applies any config changes and generates deployment manifests.
+- `kev help` - run it if you're a little lost.
+
+### init
```
Perhaps change title of this section to be more descriptive.. e.g. `Project initialisation`, `Initialise project` or similar?

ezodude replied:
@marcinc It's meant to mirror the command `init`.

I played with the other titles, but thought the simpler title would be ideal for a `Quickstart`.

ezodude replied:
embed into render.

marcinc commented on README.md:
````diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
 ```
 
-The command above will auto-detect default `docker-compose.yaml` and `docker-compose.override.yaml` files (if present) in the project directory, and tracks them as Kubernetes deployment sources.
+This identifies the default `docker-compose.yaml` and (if present) `docker-compose.override.yaml` files in your project directory. They will be used as Compose Kubernetes sources.
````
need to clarify that so it's less confusing... `as Compose Kubernetes sources` -> perhaps `as source of truth for your application deployment in Kubernetes`?

marcinc commented on README.md:
```diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
+Also, it creates a default `dev` environment and its Compose override file.  
 
-And, to individually control the configuration of your project for deployments in separate target environments, just specify their respective names with `-e` flag, e.g.:
+Here's another example. It uses an alternate `docker-compose` file with `stage` + `prod` environments:
```
`+` -> `&`

marcinc commented on README.md:
````diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
 ```
 
-Your project is now ready and environment specific configuration files generated. _Kev_ will infer useful configuration details already present in the source compose file(s). If this is not possible it'll assume sensible defaults. Users have an easy way to manually alter values directly in the environment specific override file as and when required.
+Here's what happened, _Kev_,
````
Rephrase: 

```
Here's what happened... _Kev_ has:
- Inferred the configuration details already present in your docker compose source files.
...
```

marcinc commented on README.md:
```diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
+- Assigned sensible defaults for any config it couldn't infer.
+- Created Compose overrides files for the `stage` and `prod` environments.
+
+That's it, _Kev_ is now bootstrapped and ready!
```
`That's it, _Kev_ is now bootstrapped and ready!` -> `That's it, your _Kev_ project is now bootstrapped and ready!`

marcinc commented on README.md:
```diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
+You can now customise your deployment targets by altering values in the relevant Compose environment override file. 
 
-To render manifests, simply run the following command from your project root:
+### render
```
`render` -> `Render K8s manifests`

ezodude replied:
Similar rationale to `init` as explained above.

marcinc commented on README.md:
````diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
 ```
 
-`render` generates the project's manifests based on the tracked docker-compose files, using the desired output format specified via the `-f` flag (defaults to `kubernetes`), and for selected environments. Note that ALL environments will be rendered by default if none are specified.
+The command,
````
`The command,` -> `The command above,`

marcinc commented on README.md:
```diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
 
-`render` generates the project's manifests based on the tracked docker-compose files, using the desired output format specified via the `-f` flag (defaults to `kubernetes`), and for selected environments. Note that ALL environments will be rendered by default if none are specified.
+The command,
+- Detects edits you made to the project's source compose file(s) to re-infer config changes.
```
drop ` to re-infer config changes` as it's covered by the second item already.

marcinc commented on README.md:
```diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
+- Detects edits you made to the project's source compose file(s) to re-infer config changes.
+- Applies any found config changes to your compose environment overrides.
+- Generates kubernetes manifests based on all compose files including environment overrides.
+- Targets all environments.
```
rephrase `Targets all environments` -> `Produces manifests for all environments`

marcinc commented on README.md:
````diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
-You may also specify the output directory for generated manifests with `-d` flag. Note, that specified directory will contain sub-directories, each for a separate environment name for which manifests were generated.
+```sh
+├── k8s     # stores the Kubernetes manifests for all target deployment environments. 
+├── ...
````
^ would be good to give examples including subdirectories for stages from the example above.

ezodude replied:
done

marcinc commented on README.md:
```diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
 
-## Commands
+To deploy your app onto Kubernetes,
+- Ensure you can access a running Kubernetes installation, either locally or remotely.
```
Link `locally` to something like minikube or docker-for-desktop etc?

marcinc commented on README.md:
```diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
-## Commands
+To deploy your app onto Kubernetes,
+- Ensure you can access a running Kubernetes installation, either locally or remotely.
+- Use `kubectl` to apply the manifests.
```
link `kubectl` to official page with download / installation instructions etc.

marcinc commented on README.md:
```diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
-- `kev init` - initiate project for kev.
-- `kev render` - render application manifests for selected environments, according to desired output format.
-- `kev help` - run it if you're a little lost.
+In this example, we deploy the stage environment:
```
stage -> `stage`

marcinc reviewed (commented):
initial comments to the quick start section. moving onto Tutorial.

marcinc reviewed (changes requested):
A bunch of change suggestions and comments. Let's extract tutorial to a separate .md file and link it in the README. Similar approach for other additional follow-up write-ups.

ezodude commented on README.md:
```diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
+
+_Compose_, helps you wire and run your application on a single machine with a single instance running for every service. It's super for development.
+
+_Kubernetes_, however, is designed to help you run your application in a highly available mode using clusters and service replications. It's build for production like environments.
```
I think ... `It is production grade` will suffice.

ezodude commented on README.md:
```diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
+
+Describing Compose services as Kubernetes services requires an extra layer that translates between these different worlds.
+
+Furthermore, since you're moving your app to Kubernetes, you're moving from a development mindset to sharing your app for testing, staging and at some point publishing to a production environment.
```
I see what you're saying will reword to capture the gist.

ezodude commented on README.md:
````diff
@@ -31,72 +31,585 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
+```
+
+_Kev_ has now been initialised and configured. It has,
+- Started tracking the `docker-compose.yaml` file as a Compose Kubernetes source.
````
I'm trying to establish a vocabulary to explain key concepts, hence `Compose Kubernetes source`.

I'll have a think to see it still makes sense.

ezodude replied:
Let's got with: `as the source application manifest`

... so should read:
- Started tracking the `docker-compose.yaml` file as the source application manifest.

ezodude commented on README.md:
```diff
@@ -31,72 +31,585 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
+  wordpress:
+    labels:
+      kev.service.type: LoadBalancer
+      kev.workload.cpu: "0.1"
```
I agree ... it really bugs me.

ezodude commented on README.md:
```diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
+
+Your hard-earned Docker Compose skills will help you work with Kubernetes speeding up development and iteration.   
+
+We'll set up _Kev_, iterate and deploy a [WordPress application]() onto Kubernetes.
```
Not that one, it's has the wrong image. Now linking to https://docs.docker.com/compose/wordpress/

ezodude commented on README.md:
```diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
+
+Compose and Kubernetes address different problems.
+
+_Compose_, helps you wire and run your application on a single machine with a single instance running for every service. It's super for development.
```
I get the gist, re-phrasing to make sense in the greater context.

ezodude commented on README.md:
```diff
@@ -31,72 +31,585 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
+
+**This BREAKS our running app** - to fix it we need to understand how service find-ability differs between Compose and Kubernetes.
+
+#### Make db service findable in Kubernetes
```
Going for: `Fix db service discovery`

ezodude commented on README.md:
```diff
@@ -31,72 +31,585 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
+
+It has also generated three files:
+- `kev.yaml`, a manifest that describes our _Compose Kubernetes sources_.
+- `docker-compose.kev.*.yaml`, two files to represent our _Compose environment overrides_.
```
Will leave it 🙂

ezodude commented on README.md:
````diff
@@ -31,72 +31,585 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
+```
+
+In this case, _Kev_,
+- Has re-introspected our _Compose Kubernetes source_.
````
`_Source application manifest_`

ezodude commented on README.md:
```diff
@@ -31,72 +31,585 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
+- Created `local` (useful for a testing our own machine) and `staging` (useful for a testing our a remote machine) _Compose environment overrides_.
+
+It has also generated three files:
+- `kev.yaml`, a manifest that describes our _Compose Kubernetes sources_.
```
reworded to ... `our _Source application manifest_ and _Compose environment overrides_`

marcinc reviewed (commented):
A few more comments

ezodude commented on README.md:
```diff
@@ -31,72 +31,586 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
+
+Describing Compose services as a Kubernetes application requires an extra layer that translates concepts from one to the other.
+
+Furthermore, on Kubernetes, you will share and promote your app through different stages. These stages require varying settings that should ideally be captured from the start. 
```
Thx! the above explains things more clearly.

Just rephrased slightly to work in this context:
```
Furthermore, on Kubernetes, you might also want to deploy or promote your app to different "stages", commonly known as environments. Application configuration may vary depending on the environment it is deployed to due to various infrastructure or operational constraints.

So, a good approach to managing your app configuration in different environments is a must.
```

marcinc commented on README.md:
```diff
@@ -31,72 +42,120 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
+- `kev render` - detects, applies any config changes and generates deployment manifests.
+- `kev help` - run it if you're a little lost.
+
+### init
```
tiny one: make `init` and `render` (below) bold so they stand out more.

ezodude replied:
Reverted to older suggestions. They actually work better.

marcinc commented:
Looks good, one final comment. Let's move `Contents` section below Features?

ezodude commented:
- Moved TOC under features.
- Formatted TOC to make it stand out.
- Made the Quickstart more accessible in the TOC.
- Updated the Quickstart `init` and `render` headings as per older comments (they make more sense).
- Added a tutorial synopsis in the README to help readers understand what to expect.

marcinc reviewed (approved):
LGTM
